servant-examples/auth-combinator/auth-combinator.hs

```diff
@@ -9,56 +9,39 @@
 {-# LANGUAGE TypeOperators       #-}
 {-# LANGUAGE UndecidableInstances #-}
 
-import           Data.Aeson
-import           Data.ByteString          (ByteString)
-import           Data.IORef
-import           Data.Text                (Text)
+import           Control.Monad.Trans.Except (ExceptT, throwE)
+import           Data.Aeson                 hiding ((.:))
+import           Data.ByteString            (ByteString)
+import           Data.Monoid                ((<>))
+import           Data.Text                  (Text)
 import           GHC.Generics
 import           Network.Wai
 import           Network.Wai.Handler.Warp
 import           Servant
-import           Servant.Server.Internal
-
--- Pretty much stolen/adapted from
--- https://github.com/haskell-servant/HaskellSGMeetup2015/blob/master/examples/authentication-combinator/AuthenticationCombinator.hs
-
-type DBConnection = IORef [ByteString]
-type DBLookup = DBConnection -> ByteString -> IO Bool
-
-initDB :: IO DBConnection
-initDB = newIORef ["good password"]
-
-isGoodCookie :: DBLookup
-isGoodCookie ref password = do
-  allowed <- readIORef ref
-  return (password `elem` allowed)
-
-data AuthProtected
-
-instance (HasConfigEntry config DBConnection, HasServer rest config)
-  => HasServer (AuthProtected :> rest) config where
-
-  type ServerT (AuthProtected :> rest) m = ServerT rest m
-
-  route Proxy config subserver = WithRequest $ \ request ->
-    route (Proxy :: Proxy rest) config $ addAcceptCheck subserver $ cookieCheck request
-      where
-        cookieCheck req = case lookup "Cookie" (requestHeaders req) of
-            Nothing -> return $ FailFatal err401 { errBody = "Missing auth header" }
-            Just v  -> do
-              let dbConnection = getConfigEntry config
-              authGranted <- isGoodCookie dbConnection v
-              if authGranted
-                then return $ Route ()
-                else return $ FailFatal err403 { errBody = "Invalid cookie" }
-
-type PrivateAPI = Get '[JSON] [PrivateData]
-
-type PublicAPI = Get '[JSON] [PublicData]
-
-type API = "private" :> AuthProtected :> PrivateAPI
-      :<|> PublicAPI
 
+-- | A user type that we "fetch from the database" after
+-- performing authentication
+newtype User = User { unUser :: Text }
+
+
+-- | A method that, when given a password, will return a User.
+-- This is our bespoke (and bad) authentication logic.
+lookupUser :: ByteString -> ExceptT ServantErr IO User
+lookupUser cookie =
+  if cookie == "good password"
```

Comment:
The cookie contains the password? It's such a minimal example. anyway, what I cna do instead is create a fake database in a `Map String User` or something. That might be clearer.